File: Data Structures & Algorithms/minimum-stack/submission-3.java
class MinStack {
    Stack<Integer> stack;
    Stack<Integer> min;

    public MinStack() {
        stack = new Stack<>();
        min = new Stack<>();
    }
    
    public void push(int val) {
        stack.push(val);
        min.push(min.isEmpty() ? val : Math.min(min.peek(), val));
    }
    
    public void pop() {
        if (!stack.isEmpty()) {
            stack.pop();
            min.pop();
        }
        
    }
    
    public int top() {
        if (stack.isEmpty()) return -1;
        return stack.peek();
    }
    
    public int getMin() {
        if (min.isEmpty()) return -1;
        return min.peek();
    }
}

File: Data Structures & Algorithms/validate-parentheses/submission-3.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        char = {"}":"{",")":"(","]":"[",}
        for ch in s:
            if ch not in char:
                stack.append(ch)
            else:
                if not stack or char[ch] != stack[-1]:
                    return False
                stack.pop()
        return not stack